### PR TITLE
[extractors/twitter] Add support for Twitter Spaces

### PR DESCRIPTION
**Pull request by nixxo**
### Description of your *pull request* and other information

</details>

Draft for support of Twitter Spaces download.

TODO:
- [x] Check live stream
- [x] Add various date infos
- [x] Check geo restriction if it exists
- [x] Fix Tests
- [x] Downloaded file not playable...

Adresses #1249

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))

### CODE REVIEW

File: yt_dlp/extractor/twitter.py [hunk 1]
```diff
@@ -681,6 +702,114 @@ def _real_extract(self, url):
...
+        },
+    }, {
+        # TODO: FIX TEST
+        # Twitter Space not started yet
+        'url': 'https://twitter.com/i/spaces/1YqGoAeoEOAxv',
+        'info_dict': {
+            'id': '1YqGoAeoEOAxv',
+            'ext': 'mp4',
+        },
+        'expected_warnings': ['Twitter Space not started yet'],
```
Comment (nixxo) [inline yt_dlp/extractor/twitter.py:730]: @pukkandan is it possible to do a test to verify a raise_no_format situation?

Comment (pukkandan) [reply yt_dlp/extractor/twitter.py:730]: 'params': {'ignore_no_formats_error': True}

Comment (nixxo) [reply yt_dlp/extractor/twitter.py:730]: no, that's not what I was looking for.

`ignore_no_formats_error` basically skips the `raise_no_formats` line and goes on executing the code.

I'm looking to something like `'expected_error'` to actually feed a url that I know generates an error and test that situation. I guess is not something possible to test, am I right?

**Comment by nixxo** [comment]
@pukkandan hi, the extractor should be ok now. A couple of notes:

- had to force `m3u8_protocol` to `m3u8` because `m3u8_native` produced an unplayable file. 
- the second test that is used to check if a stream is not started yet is obviously no longer valid. Should I just remove it?
- space http request uses twitter graphql and is a bit big because spaces is only available in twitter Api v2 and it required dev api key. Do you have a yt-dlp twitter account?

File: yt_dlp/extractor/twitter.py [hunk 2]
```diff
@@ -681,6 +702,121 @@ def _real_extract(self, url):
...
+            m3u8_id = 'live' if info['is_live'] else 'replay'
+
+            # force entry_protocol to m3u8 because m3u8_native is broken
+            formats = self._extract_m3u8_formats(
+                m3u8_url, media_key, 'm4a',
+                m3u8_id=m3u8_id, entry_protocol='m3u8',
+                live=info['is_live'])
+            # force audio-only
+            for i in range(len(formats)):
+                formats[i].update({
```
Comment (Lesmiscore) [inline yt_dlp/extractor/twitter.py:809]:
```suggestion
            for fmt in formats:
                fmt.update({
```

File: yt_dlp/extractor/twitter.py [hunk 3]
```diff
@@ -540,6 +555,12 @@ def get_binding_value(k):
                         'url': get_binding_value('broadcast_url'),
                         'ie_key': TwitterBroadcastIE.ie_key(),
                     })
+                elif card_name == 'audiospace':
+                    info.update({
+                        '_type': 'url',
+                        'url': 'https://twitter.com/i/spaces/' + get_binding_value('id'),
```
Comment (Lesmiscore) [inline yt_dlp/extractor/twitter.py:561]: f-string

File: yt_dlp/extractor/twitter.py [hunk 4]
```diff
@@ -681,6 +702,121 @@ def _real_extract(self, url):
...
+        formats = []
+        media_key = info.pop('media_key')
+
+        if info['live_status'] == 'is_upcoming':
+            self.raise_no_formats('Twitter Space not started yet', expected=True)
+        elif info['live_status'] == 'post_live':
+            self.raise_no_formats('Twitter Space ended but not downloadable yet', expected=True)
+        else:
+            source = self._call_api(
+                'live_video_stream/status/' + media_key, media_key)['source']
```
Comment (Lesmiscore) [inline yt_dlp/extractor/twitter.py:798]: f-string for the first parameter

File: yt_dlp/extractor/twitter.py [hunk 5]
```diff
@@ -681,6 +702,121 @@ def _real_extract(self, url):
...
+        SPACE_SATUS = {
+            'notstarted': 'is_upcoming',
+            'ended': 'was_live',
+            'running': 'is_live',
+            'timedout': 'post_live',
+        }
+        live_status = SPACE_SATUS.get(metadata.get('state').lower())
+
+        # partecipants
+        partecipants = join_nonempty(*[p.get('display_name') for p in traverse_obj(data, ('participants', 'speakers'))], delim=', ') or 'nobody yet.'
```
Comment (Lesmiscore) [inline yt_dlp/extractor/twitter.py:755]: `participants`?

Comment (Lesmiscore) [reply yt_dlp/extractor/twitter.py:758]: [So the description also has a typo?](https://en.wiktionary.org/wiki/partecipate)

Comment (nixxo) [reply yt_dlp/extractor/twitter.py]: ok, understood now that the problem was a typo. Fixed it.

**Comment by pukkandan** [comment]
Merged in https://github.com/yt-dlp/yt-dlp/commit/7a26ce2641c45b561dde190e2eb92b7d923ca5de

**Pull request closed without merging**